### ISSUE TITLE
why halide simplify & to %, << to * ?

### ISSUE THREAD

**Issue by xqdan** [halide/Halide#3079]
It seems to bit operations is more effective than % or * , why halide do it in opposite way?

https://github.com/halide/Halide/blob/master/src/Simplify.cpp#L4870
https://github.com/halide/Halide/blob/master/src/Simplify.cpp#L4904

thanks

**Comment by zvookin** [comment]
Because it provides a consistent representation with obvious mathematical
reasoning. Division and modulous by powers of two are strength reduced at
code generation time so performance should be taken care of.

-Z-

**Comment by xqdan** [comment]
Do you mean optimized results will be applied to somewhere in Halide pass? if true, could you give an example of div and mode using for mathematical reasoning?

**Comment by abadams** [comment]
x / 256 and x >> 8 both compile to x >> 8 (https://github.com/halide/Halide/blob/master/src/CodeGen_LLVM.cpp#L1309), but we want to pick a single canonical representation inside the compiler so that we don't have to implement visitors in our analysis passes that handle both ways of expressing things. E.g. I don't want to have to reason about the bounds of x/8 and x>>3 separately given some range for x - the logic would be redundant (https://github.com/halide/Halide/blob/master/src/Bounds.cpp#L393)

We found it more convenient to use div/mod as the canonical representation, because the analysis passes then also handle things like x/7. The downside of this is that the analysis passes don't do a great job of things like x & 0x010101, which cannot be converted to a modulo operation, but that sort of thing rarely shows up in indexing code.

Note that in C it's not true for signed types that x/256 and x>>8 are the same, but in Halide we defined division in a way that makes it true (round to negative infinity instead of round to zero).
